bicytok/distanceMetricFuncs.py

```diff
@@ -1,641 +1,221 @@
+                targAbunAll = np.vstack((targAbun1, targAbun2)).transpose()
+                offTargAbunAll = np.vstack((offTargAbun1, offTargAbun2)).transpose()
+
+                assert targAbunAll.shape == (np.sum(targ), 2)
+
+                KL_div_vals[rec2, rec1] = calculateKLDiv(
+                    targAbunAll, offTargAbunAll, dims=2
+                )
+
+                M = ot.dist(targAbunAll, offTargAbunAll)
+                a = np.ones((targAbunAll.shape[0],)) / targAbunAll.shape[0]
+                b = np.ones((offTargAbunAll.shape[0],)) / offTargAbunAll.shape[0]
+                EMD_vals[rec2, rec1] = ot.emd2(
+                    a, b, M, numItermax=1
+                )  # Check numIterMax
 
+            else:
+                KL_div_vals[rec2, rec1] = None
+                EMD_vals[rec2, rec1] = None
+
+    return KL_div_vals, EMD_vals
+
+
+def KL_EMD_3D(recAbundances, targ, offTarg):
     """
-    CITE_DF = importCITE()
-
-    weightDF = convFactCalc(CITE_DF)
-    exclude_columns = ["CellType1", "CellType2", "CellType3", "Cell"]
-    threshold_multiplier = 5
-    # Calculate the mean and standard deviation for each numeric column
-    numeric_columns = [col for col in dataset1.columns if col not in exclude_columns]
-    column_means = dataset1[numeric_columns].mean()
-    column_stddevs = dataset1[numeric_columns].std()
-    # Identify outliers for each numeric column
-    outliers = {}
-    for column in numeric_columns:
-        threshold = column_means[column] + threshold_multiplier * column_stddevs[column]
-        outliers[column] = dataset1[column] > threshold
-    # Create a mask to filter rows with outliers
-    outlier_mask = pd.DataFrame(outliers)
-    dataset = dataset1[~outlier_mask.any(axis=1)]
-    # receptor_names = [col for col in dataset.columns if col not in exclude_columns]
-    results = []
-    target_cells_df = dataset[
-        (dataset["CellType3"] == target_cells)
-        | (dataset["CellType2"] == target_cells)
-        | (dataset["CellType1"] == target_cells)
-    ]
-    off_target_cells_df = dataset[
-        ~(
-            (dataset["CellType3"] == target_cells)
-            | (dataset["CellType2"] == target_cells)
-            | (dataset["CellType1"] == target_cells)
-        )
-    ]
-    receptor_names = [
-        col
-        for col in dataset.columns
-        if col not in exclude_columns
-        and np.mean(target_cells_df[col]) > np.mean(off_target_cells_df[col])
-        and np.mean(target_cells_df[col]) > 5
-    ]
-
-    for receptor1_name in receptor_names:
-        for receptor2_name in receptor_names:
-            for receptor3_name in receptor_names:
+    3D applicatin of KL Divergence and EMD
+    NOT YET IMPLEMENTED
+    """
+
+    x = recAbundances
+
+    KL_div_vals = np.empty([x.shape[1], x.shape[1], x.shape[1]])
+    KL_div_vals[:] = np.nan
+    EMD_vals = np.empty([x.shape[1], x.shape[1], x.shape[1]])
+    EMD_vals[:] = np.nan
+
+    for rec1 in range(x.shape[1]):
```

Comment:
I think you can make one loop as `for i, j, k in combinations(x.shape[1],3):`. `combinations()` is in the `itertools` Python module.